spec/routes/spec_helper.rb

```diff
@@ -16,6 +16,21 @@
   def app
     Clover.freeze.app
   end
+
+  RSpec::Matchers.define :have_api_error do |expected_state, expected_message, expected_details|
+    match do |response|
+      parsed_body = JSON.parse(response.body)
+      response.status == expected_state &&
+        (expected_message.nil? || parsed_body.dig("error", "message") == expected_message) &&
+        (expected_details.nil? || parsed_body.dig("error", "details") == expected_details)
+    end
+
+    failure_message do |response|
+      parsed_body = JSON.parse(response.body)
+      "expected: ".rjust(16) + expected_state.to_s + (expected_message && " - #{expected_message}").to_s + (expected_details && " - #{expected_details}").to_s + "\n" +
```

Comment:
not sure, but maybe this would be slightly easier to read with a heredoc?

```rb
<<~MESSAGE
  #{'expected:'.rjust(16)}#{expected_state}#{expected_message ...}
  #{'got:'.rjust(16)}#{response.status}#{expected_message...}
MESSAGE
```

Reply:
Thanks for advice. heredoc makes it more readable since it removes multiple `+` and `to_s`. Changed it